sympy/printing/latex.py

```diff
@@ -1914,7 +1915,7 @@ def _print_NDimArray(self, expr):
 
         return out_str
 
-    def _printer_tensor_indices(self, name, indices, index_map={}):
+    def _printer_tensor_indices(self, name, indices, index_map={}) -> str:
```

Comment:
Why doesn't Pyright get this right? Is there any value for the result of `latex_escape(s)` that can make it not return a `str`? If there isn't, this should be reported as a bug there instead of requiring a modification in SymPy

Reply:
*same for all other similar cases

Reply:
`latex_escape` doesn't belong to those `_print_*` functions whose return type should be deduced automatically. Without hinting its parameter pyright can't deduce its return type. With hinting its parameter, pyright can, and I still add explicit type hint to it just for consistency.
`_printer_tensor_indices` is indeed weird. `index_map` has a mutable default value which is a bad practice. If modify it to `index_map: dict` then pyright can get it right.

Reply:
But regardless of the return type of `latex_escape`, the return type will be a string. I don't think `str % anything` returns anything other than a `str`?

Reply:
Which function are you referring to? There is no calls to `latex_escape` or `%` format for `_printer_tensor_indices`.

Reply:
I’m reluctant to open more issues in pyright. As you can see, most likely I’ll get an “as designed” response. It seems their culture that users should always modify their legal python code to fit their tool, otherwise just don’t use it.

Reply:
```python
class Foo(str):
    def __rmod__(self, other: object) -> int:
        return 0


print("x" % Foo())  # Gives 0
```

So yes it's possible for `str % anything` to be not a string.

Reply:
> I’m reluctant to open more issues in pyright. As you can see, most likely I’ll get an “as designed” response. It seems their culture that users should always modify their legal python code to fit their tool, otherwise just don’t use it.

That is a good point. Then maybe the problem is that `%` is used over `.format()`?

Reply:
Does pyright understand these:
```python
class Foo(str):
    def __rmod__(self, other: object) -> int:
        return 0


print("x" % Foo())  # Gives 0
print(type("{0}".format(Foo())))
print(type(f"{Foo()}"))
```
I think that fstrings are very elegant in this way.

Reply:
Thank you. This is the counter-example I expected from the pyright issue.

Reply:
> Why doesn't Pyright get this right?

Some are because the return value of a `_print_*` is a call to a helper function that is defined afterward, and I added type hints from the beginning to the end, so at the time I check the deduced return type of `_print_*` pyright cannot give correct answer.
Some are because pyright needs parameter type, although when I hover on any return value in vscodium, pyright tells me it's `str`.
Now there's only one function `_print_Function` that has to be explicitly typed. And this is obviously the reason:
https://github.com/sympy/sympy/blob/918eb81f4bb3ae062cef98a9dcb9ecda058e33d5/sympy/printing/latex.py#L940